parquet-hadoop/src/test/java/org/apache/parquet/hadoop/TestParquetWriterError.java

```diff
@@ -0,0 +1,193 @@
+import org.apache.parquet.bytes.DirectByteBufferAllocator;
+import org.apache.parquet.bytes.TrackingByteBufferAllocator;
+import org.apache.parquet.column.ParquetProperties;
+import org.apache.parquet.example.data.Group;
+import org.apache.parquet.filter2.recordlevel.PhoneBookWriter;
+import org.apache.parquet.hadoop.codec.CleanUtil;
+import org.apache.parquet.hadoop.example.ExampleParquetWriter;
+import org.apache.parquet.hadoop.metadata.CompressionCodecName;
+import org.apache.parquet.io.LocalOutputFile;
+import org.junit.Assert;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.TemporaryFolder;
+
+/**
+ * Unit test to check how Parquet writing behaves in case of an error happens during the writes. We use an OOM because
+ * that is the most tricky to handle. In this case we shall avoid flushing since it may cause writing to already
+ * released memory spaces.
+ * <p>
+ * To catch the potential issue of writing into released ByteBuffer objects, direct memory allocation is used and at the
+ * release() call we actually release the related direct memory and zero the address inside the ByteBuffer object. As a
+ * result, a subsequent read/write call on the related ByteBuffer object will crash the whole jvm. (Unfortunately, there
+ * is no better way to test this.) To avoid crashing the test executor jvm, the code of this test is executed in a
+ * separate process.
+ */
+public class TestParquetWriterError {
+
+  @Rule
+  public TemporaryFolder tmpFolder = new TemporaryFolder();
+
+  @Test
+  public void testInSeparateProcess() throws IOException, InterruptedException {
+    String outputFile = tmpFolder.newFile("out.parquet").toString();
+
+    String classpath = System.getProperty("java.class.path");
+    String javaPath = Paths.get(System.getProperty("java.home"), "bin", "java")
+        .toAbsolutePath()
+        .toString();
+    Process process = new ProcessBuilder()
+        .command(javaPath, "-cp", classpath, Main.class.getName(), outputFile)
+        .redirectError(ProcessBuilder.Redirect.INHERIT)
+        .redirectOutput(ProcessBuilder.Redirect.INHERIT)
+        .start();
+    Assert.assertEquals(
+        "Test process exited with a non-zero return code. See previous logs for details.",
+        0,
+        process.waitFor());
+  }
+
+  /**
+   * The class to be used to execute this test in a separate thread.
+   */
+  public static class Main {
+
+    private static final Random RANDOM = new Random(2024_02_27_14_20L);
+
+    // See the release() implementation in createAllocator()
+    private static final Field BUFFER_ADDRESS;
+
+    static {
+      Field bufferAddress;
+      try {
+        Class<?> bufferClass = Class.forName("java.nio.Buffer");
+        bufferAddress = bufferClass.getDeclaredField("address");
+        bufferAddress.setAccessible(true);
+      } catch (Exception e) {
+        // From java 17 it does not work, but we still test on earlier ones, so we are fine
```

Comment:
Do you mean this test will fall on java 17?

Reply:
Yes, it fails at `setAccessible(true)`.